### ISSUE TITLE
storage e2e tests: Leaking PDs due to #87258

### ISSUE THREAD

**Issue by verult** [kubernetes/kubernetes#88355]
**Which jobs are flaking**: Since #87258 was merged, storage e2e tests were leaking ~1 PD ~40% of the time. The PR has since been reverted (#87664)
See https://github.com/kubernetes/kubernetes/pull/87258#issuecomment-579932107

**Which test(s) are flaking**:

**Testgrid link**: https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=8&include-filter-by-regex=diffResources (unfortunately the failed tests are no longer in Prow, though it's still possible to find them in GCS artifacts).

**Reason for failure**: What we know so far:
* Test failure is caused by in-line volumes.
* Disk can’t be deleted in the end because it’s in use (presumably not detached successfully?)
* But the pod using the volume was deleted already (supposedly, given log messages), and very quickly.
* Given this is a flaky behavior, it’s likely caused by a race (maybe in NestedPendingOperations?)
* This test can be run in parallel (and probably is parallel in the test failures)

/assign

**Comment by verult** [comment]
One of the failed runs (at Wed, 29 Jan 2020 18:28:59 GMT): https://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/1222575922826186752/

**Comment by verult** [comment]
Discovered that `getOperation()` and `deleteOperation()` aren't checking `nodeName`.

https://github.com/verult/kubernetes/blob/c6a03fa5be5a3a8eaf1d3953918390230d8115ef/pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go#L225-L230

https://github.com/verult/kubernetes/blob/c6a03fa5be5a3a8eaf1d3953918390230d8115ef/pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go#L239-L245

Need to verify whether this could lead to the test failure.

Next steps:
* Write a unit test to capture the failure scenario and try to repro.
* Try to repro in CI with a dummy PR (local repro was unsuccessful).
* Fix `getOperation()` and `deleteOperation()` and add tests

**Comment by verult** [comment]
Fixed PR: #88678

I was able to repro the failure in a test cluster and confirmed the leak is fixed across 10 runs. The PR also includes a unit test that captures a scenario that likely has caused the failure, and also captures the fix.

To increase test coverage, I considered building a fuzzer for `NestedPendingOperations`, introducing many operations with different timing to capture races, but will look into e2e stress tests instead since the fuzzer may test inputs and timings that may not actually be exercised.

Next steps:
* Look into adding stress tests to cover race conditions.
* Refactor operation keys

**Comment by verult** [comment]
Some details on the flaky test failure: example sequence of AttachDetachController logs for a leaked disk:

```
I0129 18:01:20.902965       1 reconciler.go:272] attacherDetacher.AttachVolume started for volume "gcepd-volume-0" (UniqueName: "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a") from node "bootstrap-e2e-minion-group-fx7r" 
I0129 18:01:25.160498       1 operation_generator.go:359] AttachVolume.Attach succeeded for volume "gcepd-volume-0" (UniqueName: "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a") from node "bootstrap-e2e-minion-group-fx7r" 
I0129 18:01:25.160526       1 actual_state_of_world.go:479] Report volume "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a" as attached to node "bootstrap-e2e-minion-group-fx7r"
I0129 18:02:00.638226       1 actual_state_of_world.go:399] Set detach request time to current time for volume kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a on node "bootstrap-e2e-minion-group-fx7r"
I0129 18:02:00.733784       1 reconciler.go:200] attacherDetacher.DetachVolume started for volume "gcepd-volume-0" (UniqueName: "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a") on node "bootstrap-e2e-minion-group-fx7r" 
I0129 18:02:00.769063       1 operation_generator.go:1346] Verified volume is safe to detach for volume "gcepd-volume-0" (UniqueName: "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a") on node "bootstrap-e2e-minion-group-fx7r" 
I0129 18:02:01.341211       1 reconciler.go:272] attacherDetacher.AttachVolume started for volume "gcepd-volume-0" (UniqueName: "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a") from node "bootstrap-e2e-minion-group-fkjx" 
I0129 18:02:05.663404       1 gen.go:13551] GCEInstances.AttachDisk(context.Background.WithDeadline(2020-01-29 19:02:01.78025169 +0000 UTC m=+4267.754879467 [59m56.11683667s]), Key{"bootstrap-e2e-minion-group-fkjx", zone: "us-west1-b"}, ...) = googleapi: Error 400: RESOURCE_IN_USE_BY_ANOTHER_RESOURCE - The disk resource 'projects/k8s-jkns-e2e-gce-1-6/zones/us-west1-b/disks/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a' is already being used by 'projects/k8s-jkns-e2e-gce-1-6/zones/us-west1-b/instances/bootstrap-e2e-minion-group-fx7r'
E0129 18:02:05.663484       1 attacher.go:106] Error attaching PD "bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a" to node "bootstrap-e2e-minion-group-fkjx": googleapi: Error 400: RESOURCE_IN_USE_BY_ANOTHER_RESOURCE - The disk resource 'projects/k8s-jkns-e2e-gce-1-6/zones/us-west1-b/disks/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a' is already being used by 'projects/k8s-jkns-e2e-gce-1-6/zones/us-west1-b/instances/bootstrap-e2e-minion-group-fx7r'
I0129 18:02:05.663536       1 actual_state_of_world.go:322] Volume "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a" is already added to attachedVolume list to node "bootstrap-e2e-minion-group-fkjx", update device path ""
E0129 18:02:05.663625       1 nestedpendingoperations.go:285] Operation for "{volumeName=\"kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a\", podName=\"\", nodeName=\"bootstrap-e2e-minion-group-fkjx\"}" failed. No retries permitted until 2020-01-29 18:02:06.163596233 +0000 UTC m=+672.138224024 (durationBeforeRetry 500ms). Error: "AttachVolume.Attach failed for volume \"gcepd-volume-0\" (UniqueName: \"kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a\") from node \"bootstrap-e2e-minion-group-fkjx\" : googleapi: Error 400: RESOURCE_IN_USE_BY_ANOTHER_RESOURCE - The disk resource 'projects/k8s-jkns-e2e-gce-1-6/zones/us-west1-b/disks/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a' is already being used by 'projects/k8s-jkns-e2e-gce-1-6/zones/us-west1-b/instances/bootstrap-e2e-minion-group-fx7r'"
I0129 18:02:05.664158       1 event.go:278] Event(v1.ObjectReference{Kind:"Pod", Namespace:"volume-3037", Name:"gcepd-client", UID:"feba1973-1fbc-4c60-9381-b1effef5cf60", APIVersion:"v1", ResourceVersion:"16985", FieldPath:""}): type: 'Warning' reason: 'FailedAttachVolume' AttachVolume.Attach failed for volume "gcepd-volume-0" : googleapi: Error 400: RESOURCE_IN_USE_BY_ANOTHER_RESOURCE - The disk resource 'projects/k8s-jkns-e2e-gce-1-6/zones/us-west1-b/disks/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a' is already being used by 'projects/k8s-jkns-e2e-gce-1-6/zones/us-west1-b/instances/bootstrap-e2e-minion-group-fx7r'
I0129 18:02:06.050096       1 operation_generator.go:470] DetachVolume.Detach succeeded for volume "gcepd-volume-0" (UniqueName: "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a") on node "bootstrap-e2e-minion-group-fx7r" 
I0129 18:02:06.096513       1 reconciler.go:272] attacherDetacher.AttachVolume started for volume "gcepd-volume-0" (UniqueName: "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a") from node "bootstrap-e2e-minion-group-fkjx" 
I0129 18:02:10.618685       1 operation_generator.go:359] AttachVolume.Attach succeeded for volume "gcepd-volume-0" (UniqueName: "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a") from node "bootstrap-e2e-minion-group-fkjx" 
I0129 18:02:10.618725       1 actual_state_of_world.go:322] Volume "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a" is already added to attachedVolume list to node "bootstrap-e2e-minion-group-fkjx", update device path "/dev/disk/by-id/google-bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a"
I0129 18:02:10.618739       1 actual_state_of_world.go:479] Report volume "kubernetes.io/gce-pd/bootstrap-e2e-5164f7d2-4bc8-45ff-88ba-6654dd9f024a" as attached to node "bootstrap-e2e-minion-group-fkjx"
```

In other words:
1. Attach(volume, node1) start
1. Attach(volume, node1) success
1. Detach(volume, node1) start
1. Attach(volume, node2) start
1. Attach(volume, node2) failure because volume is still attached to node1. Operation enters exponential backoff
1. Detach(volume, node1) success
1. Attach(volume, node2) start (retry from previous op)
1. Attach(volume, node2) success

And then silence. In a correct sequence, there should be a Detach(volume, node2) at the very end. The pod was deleted at the end of the test, but detach never happened.

The bug that `getOperation()` and `deleteOperation()` aren't checking nodeName introduced a whole host of possible race conditions. My closest theory (which has 1 inconsistency with the log output) is the following:

1. (skipping Attach(volume, node1) as it doesn't play a role in introducing the race condition proposed here. We assume its operations are entirely correct)
1. Attach(someOtherVolume) start. `NestedPendingOperation` keys: `[(someOtherVolume)]`
  * The in-line volume test is run in parallel with other storage tests, so this is possible. The exact key doesn't matter (i.e. could even have a nodeName).
1. Detach(volume, node1) start. `NestedPendingOperation` keys: `[(someOtherVolume), (volume, node1)]`
  * The order of keys matter.
1. Attach(volume, node2) start. `NestedPendingOperation` keys: `[(someOtherVolume), (volume, node1), (volume, node2)]`
1. Attach(someOtherVolume) succeeds. `[(volume, node2), (volume, node1)]`
  * Due to the way `deleteOperation()` works, `(volume, node2)` is moved to the front.
1. Attach(volume, node2) failure because volume is still attached to node1. Operation enters exponential backoff. `(volume, node2)` is updated.
1. Detach(volume, node1) success. `[(volume, node1)]`
  * Due to the bug in `deleteOperation()`, the key with the first volume match is deleted. Here the deletion is wrong.
1. Attach(volume, node2) start (retry from previous op).  `[(volume, node1), (volume, node2)]`
1. Attach(volume, node2) success. `[(volume, node2)]`
  * The wrong key is deleted again, leaving a dangling `node2` operation.

Now calls to `Detach(volume, node2)` will permanently result in an `alreadyExistsError`. The inconsistency with the test is failure is that there should have been an error log printed by the ADC renconciler, but there wasn't. On the other hand, it couldn't have been stuck on exponential backoff either because max backoff duration is 2min2sec, while the test retried PD deletion for 5min. Also, in my repro leaked disks are never detached even when I leave the cluster running for a while. 

So other than the bug fixed in #88678, there may be something else at play here and I'd like to dig deeper. But #88678 addressed a major error causing race conditions and stopped the leak in my repros, so it's still worth getting it in before 1.18 code freeze.

The case described above is also captured in a unit test in https://github.com/kubernetes/kubernetes/pull/88678.

**Comment by verult** [comment]
This has been fixed by #88678
